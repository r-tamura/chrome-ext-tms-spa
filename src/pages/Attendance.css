.attendance-holiday {
  color: rgba(255, 0, 0, .4);
  font-weight: var(--font-bold);
}
