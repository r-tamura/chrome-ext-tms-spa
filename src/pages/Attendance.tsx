import * as React from "react";
import { connect } from "react-redux";
import { Helmet } from "react-helmet";
import { RouteComponentProps } from "react-router-dom";
import SelectBox from "~/components/SelectBox";
import { Button } from "~/components/Button";
import {
  RootState,
  getProjects,
  getAttendancesSelectedMonth,
  getAttendanceSettings
} from "~/modules";
import {
  fetchAttendancesIfNeeded,
  saveAttendancesIfNeeded,
  updateDaily,
  fetchSettings,
  setMonthlyWithDefaults,
  updateSettings,
  changeMonth
} from "~/modules/attendances";
import { zerofill } from "~/helpers/common";
import { AttendanceMonthlyView, AttendanceSettings, Project } from "~/types";
import { MsgAttendance } from "~/helpers/_const";

interface OwnProps extends RouteComponentProps<{}>, React.Props<{}> {}

interface IProps extends OwnProps {
  attendanceMonthly: AttendanceMonthlyView;
  attendanceSettings: AttendanceSettings;
  projects: Project[];
  fetchAttendancesIfNeeded: () => any;
  fetchSettings: () => any;
  saveAttendancesIfNeeded: () => any;
  updateDaily: (dailyId: string, patch: any) => any;
  setMonthlyWithDefaults: (dailyIds: string[]) => any;
  updateSettings: (patch: Partial<AttendanceSettings>) => any;
  changeMonth: (year: number, month: number) => any;
  submitApplication: (year: number, month: number) => any;
}

class AttendancePage extends React.Component<IProps, {}> {
  constructor(props: IProps) {
    super(props);
    this.props.fetchSettings();
    this.props.fetchAttendancesIfNeeded();
  }

  public render() {
    const { attendanceMonthly } = this.props;
    const { year, month, days } = attendanceMonthly;
    return (
      <main className="main">
        <Helmet>
          <title>Attendance | TMS</title>
        </Helmet>
        <div className="main-column">
          <h1>
            <Button id={"btn-prev-month"} onClick={this.onPrevMonthClick}>
              {"<"}
            </Button>
            {year} / {month}
            <Button id={"btn-next-month"} onClick={this.onNextMonthClick}>
              {">"}
            </Button>
          </h1>
          {this.renderMonthly(attendanceMonthly)}

          {/* 時間リスト */}
          <datalist id="timelist-am">
            <option>09:00</option>
            <option>09:30</option>
            <option>10:00</option>
            <option>10:30</option>
          </datalist>
          <datalist id="timelist-pm">
            <option>17:30</option>
            <option>18:00</option>
            <option>18:30</option>
            <option>19:00</option>
            <option>19:30</option>
            <option>19:30</option>
            <option>20:00</option>
            <option>20:30</option>
            <option>21:00</option>
            <option>21:30</option>
            <option>22:00</option>
            <option>22:30</option>
            <option>23:00</option>
            <option>23:30</option>
          </datalist>
        </div>
      </main>
    );
  }

  private renderMonthly(attendanceMonthly: AttendanceMonthlyView) {
    const { year, month, days, isFetching, hasApplied } = attendanceMonthly;

    if (isFetching) {
      // TODO: ロードインジケータを表示するように
      return <p>Now Loading...</p>;
    }

    if (days.length === 0) {
      return <p>勤怠データを取得できません</p>;
    }

    return (
      <div>
        <div className={"tms-btn-group"}>
          <Button id={"btn-set-default"} onClick={this.onSetDefaultClick}>
            set default
          </Button>
          <Button id={"btn-save"} onClick={this.onSaveClick}>
            Save
          </Button>
          <Button
            id={"btn-fetch"}
            onClick={this.props.fetchAttendancesIfNeeded}
          >
            Reload
          </Button>
          {hasApplied ? (
            <Button id={"btn-submit"} title={"上長申請中"} disabled>
              Submit
            </Button>
          ) : (
            <Button
              id={"btn-submit"}
              title={"上長申請"}
              onClick={this.onSubmit}
            >
              Submit
            </Button>
          )}
        </div>
        <div>
          <SelectBox
            value={this.props.attendanceSettings.projectId || ""}
            options={{
              items: this.props.projects,
              valueKey: "projectId",
              labelKey: "name"
            }}
            onChange={$select =>
              this.props.updateSettings({
                projectId: $select.selectedOptions[0].value
              })
            }
          />
        </div>
        <table className="tms-table tms-table--bordered">
          <thead>
            <tr>
              <th>日</th>
              <th>出勤</th>
              <th>退勤</th>
              {/* <th>合計</th>
              <th>Extra</th>
              <th>MN Extra</th> */}
              <th>プロジェクト</th>
              <th>承認</th>
            </tr>
          </thead>
          <tbody>
            {days.map((a, index) => (
              <tr
                key={a.dailyId}
                className={a.hasUpdated ? "attendance-diff" : ""}
              >
                <td
                  className={
                    a.isWeekday ? "attendance-weekday" : "attendance-holiday"
                  }
                >
                  {zerofill(2, index + 1)}
                </td>
                <td className="tms-textfield tms-textfield--table">
                  <input
                    type="time"
                    value={a.start}
                    list="timelist-am"
                    onChange={e =>
                      this.props.updateDaily(a.dailyId, {
                        start: e.target.value
                      })
                    }
                  />
                </td>
                <td className="tms-textfield tms-textfield--table">
                  <input
                    type="time"
                    value={a.end}
                    list="timelist-pm"
                    onChange={e =>
                      this.props.updateDaily(a.dailyId, { end: e.target.value })
                    }
                  />
                </td>
                {/* プロジェクトが未設定の場合は表示なし */}
                <td>
                  <SelectBox
                    value={a.project ? a.project.projectId : ""}
                    options={{
                      items: this.props.projects,
                      valueKey: "projectId",
                      labelKey: "name"
                    }}
                    onChange={$select =>
                      this.props.updateDaily(a.dailyId, {
                        projectId: $select.selectedOptions[0].value
                      })
                    }
                    additionalClass={"tms-select--table"}
                  />
                </td>
                <td>
                  {a.hasConfirmed
                    ? MsgAttendance.HAS_SUBMITED
                    : MsgAttendance.NOT_SUBMITED}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }

  private onSaveClick = (e: React.MouseEvent<HTMLButtonElement>): void => {
    this.props.saveAttendancesIfNeeded();
  };

  private onSetDefaultClick = (
    e: React.MouseEvent<HTMLButtonElement>
  ): void => {
    this.props.setMonthlyWithDefaults(
      this.props.attendanceMonthly.days.map(d => d.dailyId)
    );
  };

  private onNextMonthClick = (e: React.MouseEvent<HTMLButtonElement>): void => {
    const { year: curYear, month: curMonth } = this.props.attendanceMonthly;
    const nextMonth = (curMonth % 12) + 1;
    const nextYear = nextMonth === 1 ? curYear + 1 : curYear;
    this.props.changeMonth(nextYear, nextMonth);
  };

  private onPrevMonthClick = (e: React.MouseEvent<HTMLButtonElement>): void => {
    const { year: curYear, month: curMonth } = this.props.attendanceMonthly;
    const prevMonth = ((curMonth + 10) % 12) + 1;
    const prevYear = prevMonth === 12 ? curYear - 1 : curYear;
    this.props.changeMonth(prevYear, prevMonth);
  };

  private onSubmit = (e: React.MouseEvent<HTMLButtonElement>): void => {
    console.log("[Attendance] Submitted");
  };
}

const mapStateToProps = (state: RootState, onwProps: OwnProps) => {
  return {
    projects: getProjects(state),
    attendanceSettings: getAttendanceSettings(state),
    attendanceMonthly: getAttendancesSelectedMonth(state)
  };
};

export { AttendancePage };

export default connect(
  mapStateToProps,
  {
    fetchAttendancesIfNeeded,
    saveAttendancesIfNeeded,
    updateDaily,
    fetchSettings,
    setMonthlyWithDefaults,
    updateSettings,
    changeMonth
  }
)(AttendancePage);
