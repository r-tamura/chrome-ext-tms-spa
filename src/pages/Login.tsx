import * as React from "react"
import { Helmet } from "react-helmet"
import { RouteComponentProps, withRouter } from "react-router-dom"
import { RootState } from "~/modules"
import { loginUser } from "~/modules/user"
import { Dispatch } from "redux"
import { connect } from "react-redux"

interface IProps extends React.Props<{}>, RouteComponentProps<{}> {
  readonly isFetching: boolean,
  readonly isAuthenticated: boolean,
  readonly name: string,
  loginUser: (id: string, pw: string) => (dispatch: Dispatch<{}>) => any
}

/**
 * ログインセクション
 */
class Login extends React.Component<IProps, {}> {
  private inputPassword: HTMLInputElement
  private inputUserName: HTMLInputElement

  // private loginSubscribtion: { remove: Function }

  // private inputUserName: HTMLInputElement // ユーザ名テキストボックス
  // private inputPassword: HTMLInputElement // ユーザパスワードテキストボックス

  /* ログインアクション実行 */
  // componentDidMount() {
  //   // 共通クリックイベントの登録
  //   UI.addActivateEventListener(document.getElementById("login-section"))
  // }

  // componentWillUnmount() {
  //   // 共通クリックイベントの登録
  //   UI.removeActivateEventListener(document.getElementById("login-section"))
  // }

  public render() {
    const { isFetching } = this.props
    if (isFetching) {
      return <p>Fetching...</p>
    }

    return (
      <div className="login-main">
        <Helmet>
          <title>Login | TMS</title>
        </Helmet>
        <h1>Sign in to TMS</h1>
        <p>Enter your <strong>user name</strong> and <strong>password</strong>.</p>
        <form onSubmit={this.handleSubmit}>
          <div className="span-10-of-12">
              {/* User name text filed */}
              <div className="text-field">
                <input
                  ref={input => this.inputUserName = input}
                  type="text"
                  placeholder="John Doe"
                  maxLength={30}
                  required={true}
                />
                <label>UserName</label>
              </div>
              {/* Password text field */}
              <div className="text-field">
                <input
                  ref={input => this.inputPassword = input}
                  type="password"
                  placeholder="Your password"
                  maxLength={30}
                  required={true}
                />
                <label>Password</label>
              </div>
              <button className="secondary-button btn-raised span-12-of-12">Log in</button>
            </div>
        </form>
      </div>
    )
  }

  private handleSubmit = (e: React.SyntheticEvent<HTMLFormElement>) => {
    const username = this.inputUserName.value
    const password = this.inputPassword.value

    e.preventDefault()
    this.props.loginUser(username, password)
  }
}

const mapStateToProps = (state: RootState, props: IProps) => {
  const { isFetching, isAuthenticated, name } = state.user
  return {
    isFetching,
    isAuthenticated,
    name,
  }
}

const mapDispatchToProps = {
  loginUser,
}

// export default withNav<RouteComponentProps<{}>>(Login)
export default connect(mapStateToProps, mapDispatchToProps)(Login)
