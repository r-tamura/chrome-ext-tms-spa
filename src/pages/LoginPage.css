.login-main {
  width: 95%;
  align-self: center;
  justify-self: center;

  & > * {
    text-align: center;
  }

  @media (--sm) {
    width: 550px;
  }
}

