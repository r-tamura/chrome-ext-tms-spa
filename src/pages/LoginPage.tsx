import * as React from "react"
import { Helmet } from "react-helmet"
import { RouteComponentProps } from "react-router-dom"
import { Form, Text } from "react-form"
import { RootState } from "~/modules"
import { loginUser, navigateToDashBoard } from "~/modules/user"
import { Dispatch } from "redux"
import { connect } from "react-redux"

interface IProps extends React.Props<{}>, RouteComponentProps<{}> {
  readonly isFetching: boolean,
  /* ログイン済みであるか */
  readonly isAuthenticated: boolean,
  /* ユーザー名 */
  readonly name: string,
  loginUser: (id: string, pw: string) => any
  navigateToDashBoard: () => any
}

/**
 * ログインページ
 */
class LoginPage extends React.Component<IProps, {}> {

  public componentWillMount() {
    this.redirectToDashboardIfNeeded(this.props.isAuthenticated)
  }

  public render() {
    const { isFetching } = this.props
    if (isFetching) {
      return <p>Fetching...</p>
    }

    return (
      <main className="main main--full tms-grid-3-container">
        <Helmet>
          <title>Login | TMS</title>
        </Helmet>
        <div className="login-main tms-grid--offset1 tms-grid--col1 tms-grid-sm--col3">
          <h1>Sign in to TMS</h1>
          <p>Enter your <strong>user name</strong> and <strong>password</strong>.</p>
          <Form onSubmit={this.handleSubmit}>
            { formApi => (
              <div className="tms-panel">
                <form onSubmit={formApi.submitForm}>
                  {/* User name text filed */}
                  <div className="tms-textfield">
                    <Text
                      field={"username"}
                      id={"username"}
                      placeholder="John Doe"
                      maxLength={30}
                      required={true}
                    />
                    <label htmlFor={"username"}>UserName</label>
                  </div>
                  {/* Password text field */}
                  <div className="tms-textfield">
                    <Text
                      type="password"
                      field={"password"}
                      id={"password"}
                      placeholder="Your password"
                      maxLength={30}
                      required={true}
                    />
                    <label htmlFor={"password"}>Password</label>
                  </div>
                  <button className="tms-btn primary block">SIGN IN</button>
                </form>
              </div>
            )}
          </Form>
        </div>
      </main>
    )
  }

  private shouldRedirectToDashBoard(isAuthenticated: boolean): boolean {
    return isAuthenticated
  }

  private redirectToDashboardIfNeeded(isAuthenticated: boolean) {
    if (this.shouldRedirectToDashBoard(isAuthenticated)) {
      this.props.navigateToDashBoard()
    }
  }

  private handleSubmit = ({username = "", password = ""}) => {
    this.props.loginUser(username, password)
  }
}

const mapStateToProps = (state: RootState, props: IProps) => {
  const { isFetching, isAuthenticated, name } = state.user
  return {
    isFetching,
    isAuthenticated,
    name,
  }
}

const mapDispatchToProps = {
  loginUser,
  navigateToDashBoard,
}

export {
  LoginPage,
}

// export default withNav<RouteComponentProps<{}>>(Login)
export default connect(mapStateToProps, mapDispatchToProps)(LoginPage)
