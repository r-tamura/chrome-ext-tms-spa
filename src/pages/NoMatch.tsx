import * as React from "react"
import { RouteComponentProps } from "react-router-dom"

import { RootState } from "~/modules"

class NoMatch extends React.Component<RouteComponentProps<{}>, {}> {
  public render() {
    return (
    <main className="main">
      <h1>Not Found 404</h1>
    </main>
    )
  }
}

export default NoMatch
