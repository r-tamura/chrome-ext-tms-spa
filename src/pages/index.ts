export * from "./Attendance";
export * from "./SignInPage";
export * from "./TransportExpensePage";
export * from "./App";
export * from "./NoMatch";
