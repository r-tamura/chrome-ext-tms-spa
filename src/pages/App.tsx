import * as React from "react"
import { Dispatch } from "redux"
import { Helmet } from "react-helmet"
import { connect, MapStateToProps, DispatchProp } from "react-redux"
import { Route, RouteComponentProps, Switch, withRouter } from "react-router-dom"
import Header from "~/components/Header"
import Nav from "~/components/Nav"
import { RootState } from "~/modules"
import EnsureLoggedInContainer from "~/containers/EnsureLoggedInContainer"
import LoginPage from "./LoginPage"
import TransExpense from "./TransExpenses"
import Attendance from "./Attendance"
import NoMatch from "./NoMatch"

type OwnProps = IProps

interface IProps extends React.Props<{}> {
  // fetchMaster: () => Promise<{}>
}

class App extends React.Component<IProps, {}> {
  public render() {
    return (
      <div className="app-root">
        <Helmet>
          {/* iOS9のsafariではinitial-scaleが無視される場合がある */}
          <meta name="viewport" content="width=device-width, initial-scale=1.0, shrink-to-fit=no" />
        </Helmet>
        <Header />
        <Route
          path="/:page"
          render={renderProps => {
            return <Nav path={renderProps.match.url} />
          }}
        />
        <Switch>
          <Route exact={true} path="/" component={LoginPage}/>
          <EnsureLoggedInContainer>
            <Route path="/transportation" component={TransExpense}/>
            <Route path="/attendance" component={Attendance}/>
          </EnsureLoggedInContainer>
          <Route component={NoMatch}/>
        </Switch>
      </div>
    )
  }
}

// export default connect((state: RootState, ownProps: OwnProps) => ({}), { fetchMaster })(App)
export default App
