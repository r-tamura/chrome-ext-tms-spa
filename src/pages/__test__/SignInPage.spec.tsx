import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { SignInPage } from "~/pages/SignInPage";
import * as hooks from "~/stores/hooks";

// Jest Mock in TypeScript
// https://stackoverflow.com/questions/51495473/typescript-and-jest-avoiding-type-errors-on-mocked-functions
jest.mock("../../stores/hooks");
const mockedHooks = hooks as jest.Mocked<typeof hooks>;

describe("<SignInPage />", () => {
  it("should render 'Fetching...'", () => {
    mockedHooks.useUser.mockReturnValue({
      isAuthenticated: false,
      isFetching: true,
      name: null,
      login: jest.fn(),
      logout: jest.fn()
    });
    mockedHooks.useRouter.mockReturnValue({
      pathname: "/signin",
      goto: jest.fn(),
      gotoSignIn: jest.fn(),
      gotoDashboard: jest.fn(),
      gotoTransportation: jest.fn(),
      gotoAttendance: jest.fn()
    });
    const page = renderer.create(<SignInPage />).toJSON();
    expect(page).toMatchSnapshot();
  });
  it("should render Login page", () => {
    mockedHooks.useUser.mockReturnValue({
      isAuthenticated: true,
      isFetching: false,
      name: "testuser",
      login: jest.fn(),
      logout: jest.fn()
    });
    const component = mount(<SignInPage />);
    component.find("form").simulate("submit", { preventDefault: () => ({}) });
    expect(toJson(component)).toMatchSnapshot();
  });
  it("should redirect to DashBoard", () => {
    const mockedGoto = jest.fn();
    mockedHooks.useUser.mockReturnValue({
      isAuthenticated: true,
      isFetching: false,
      name: "testuser",
      login: jest.fn(),
      logout: jest.fn()
    });
    mockedHooks.useRouter.mockReturnValue({
      pathname: "/signin",
      goto: mockedGoto,
      gotoSignIn: jest.fn(),
      gotoDashboard: jest.fn(),
      gotoTransportation: jest.fn(),
      gotoAttendance: jest.fn()
    });
    mount(<SignInPage />);
    expect(mockedGoto).toBeCalled();
  });
});
