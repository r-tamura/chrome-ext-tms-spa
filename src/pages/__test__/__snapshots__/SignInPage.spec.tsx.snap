// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignInPage /> should render 'Fetching...' 1`] = `
<p>
  Fetching...
</p>
`;

exports[`<SignInPage /> should render Login page 1`] = `
<Component>
  <styled.main
    className="app-content-full"
  >
    <StyledComponent
      className="app-content-full"
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "lastClassName": "flTsVq",
            "rules": Array [
              "
  align-self: center;
  justify-self: center;
",
            ],
          },
          "displayName": "styled.main",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "main",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <main
        className="app-content-full sc-htpNat flTsVq"
      >
        <HelmetWrapper
          defer={true}
          encodeSpecialCharacters={true}
        >
          <SideEffect(NullComponent)
            defer={true}
            encodeSpecialCharacters={true}
            title="Sign In | TMS"
            titleAttributes={Object {}}
          >
            <NullComponent
              defer={true}
              encodeSpecialCharacters={true}
              title="Sign In | TMS"
              titleAttributes={Object {}}
            />
          </SideEffect(NullComponent)>
        </HelmetWrapper>
        <styled.div>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bxivhb",
                  "isStatic": true,
                  "lastClassName": "dmfQgN",
                  "rules": Array [
                    "
  width: 95%;
  align-self: center;
  justify-self: center;

  & > * {
    text-align: center;
  }

  ",
                    "@media (min-width: 768px)",
                    " {
    width: 550px;
  }
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bxivhb",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="sc-bxivhb dmfQgN"
            >
              <h1>
                Sign in to TMS
              </h1>
              <p>
                Enter your 
                <strong>
                  user name
                </strong>
                 and 
                <strong>
                  password
                </strong>
                .
              </p>
              <ReactForm
                onSubmit={[Function]}
              >
                <Connect(Form)
                  onSubmit={[Function]}
                  store={
                    Object {
                      "dispatch": [Function],
                      "getState": [Function],
                      "replaceReducer": [Function],
                      "subscribe": [Function],
                      Symbol(observable): [Function],
                    }
                  }
                >
                  <Form
                    defaultValues={Object {}}
                    dispatch={[Function]}
                    formState={
                      Object {
                        "asyncErrors": undefined,
                        "asyncSuccesses": undefined,
                        "asyncValidations": 0,
                        "asyncWarnings": undefined,
                        "errors": undefined,
                        "submits": 1,
                        "submitted": false,
                        "submitting": true,
                        "successes": undefined,
                        "touched": Object {},
                        "validating": undefined,
                        "validationFailed": undefined,
                        "validationFailures": 0,
                        "values": Object {},
                        "warnings": undefined,
                      }
                    }
                    onSubmit={[Function]}
                    preventDefault={true}
                    pure={true}
                    store={
                      Object {
                        "dispatch": [Function],
                        "getState": [Function],
                        "replaceReducer": [Function],
                        "subscribe": [Function],
                        Symbol(observable): [Function],
                      }
                    }
                    storeSubscription={
                      Subscription {
                        "listeners": Object {
                          "clear": [Function],
                          "get": [Function],
                          "notify": [Function],
                          "subscribe": [Function],
                        },
                        "onStateChange": [Function],
                        "parentSub": undefined,
                        "store": Object {
                          "dispatch": [Function],
                          "getState": [Function],
                          "replaceReducer": [Function],
                          "subscribe": [Function],
                          Symbol(observable): [Function],
                        },
                        "unsubscribe": [Function],
                      }
                    }
                  >
                    <styled.div>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bdVaJa",
                              "isStatic": false,
                              "lastClassName": "fcklsL",
                              "rules": Array [
                                "
  padding: 15px;
  margin-bottom: 20px;
  border-radius: 0;
  background-color: ",
                                [Function],
                                ";
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.16), 0 0 2px 0 rgba(0, 0, 0, 0.12);
",
                              ],
                            },
                            "displayName": "styled.div",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bdVaJa",
                            "target": "div",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <div
                          className="sc-bdVaJa fcklsL"
                        >
                          <form
                            onSubmit={[Function]}
                          >
                            <styled.div>
                              <StyledComponent
                                forwardedComponent={
                                  Object {
                                    "$$typeof": Symbol(react.forward_ref),
                                    "attrs": Array [],
                                    "componentStyle": ComponentStyle {
                                      "componentId": "sc-bwzfXH",
                                      "isStatic": false,
                                      "lastClassName": "fsDvwN",
                                      "rules": Array [
                                        "
  display: block;
  padding-top: ",
                                        "21.25px",
                                        ";
  margin-bottom: ",
                                        "20px",
                                        ";
  position: relative;

  & input,
  & textarea {
    display: block;
    background-color: transparent;
    color: ",
                                        "rgba(0, 0, 0, 0.87)",
                                        ";
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.26);
    outline: 0;
    width: 100%;
    padding: 0;
    box-shadow: none;
    border-radius: 0;
    font-size: ",
                                        "17px",
                                        ";
    font-family: inherit;
    line-height: inherit;
    background-image: none;
    text-indent: 0.5em;
  }

  & > label {
    position: absolute;
    top: 0;
    display: block;
    width: 100%;
    color: ",
                                        "rgba(0, 0, 0, 0.54)",
                                        ";
    font-size: 12px;
    font-weight: 400;
    line-height: 15px;
    overflow-x: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    text-align: left;
  }

  & input[type=\\"time\\"] {
    cursor: pointer;
  }

  /*
   * autocomplete styles in webkit browsers
   * https://css-tricks.com/snippets/css/change-autocomplete-styles-webkit-browsers/
  */
  & input:-webkit-autofill,
  & input:-webkit-autofill:hover,
  & input:-webkit-autofill:focus,
  & input:-webkit-autofill:active {
    box-shadow: 0 0 0px 10px ",
                                        [Function],
                                        "
      inset;
    -webkit-text-fill-color: ",
                                        [Function],
                                        ";
  }

  &.tms-textfield--table input {
    border-bottom: none;
  }
",
                                      ],
                                    },
                                    "displayName": "styled.div",
                                    "foldedComponentIds": Array [],
                                    "render": [Function],
                                    "styledComponentId": "sc-bwzfXH",
                                    "target": "div",
                                    "toString": [Function],
                                    "warnTooManyClasses": [Function],
                                    "withComponent": [Function],
                                  }
                                }
                                forwardedRef={null}
                              >
                                <div
                                  className="sc-bwzfXH fsDvwN"
                                >
                                  <ConnectedField
                                    field="username"
                                    id="username"
                                    maxLength={30}
                                    placeholder="John Doe"
                                    required={true}
                                  >
                                    <Field
                                      component={[Function]}
                                      field="username"
                                      id="username"
                                      maxLength={30}
                                      placeholder="John Doe"
                                      pure={true}
                                      required={true}
                                    >
                                      <TextWrapper
                                        fieldApi={
                                          Object {
                                            "addValue": [Function],
                                            "asyncValidate": [Function],
                                            "doneValidatingField": [Function],
                                            "error": undefined,
                                            "fieldName": "username",
                                            "preValidate": [Function],
                                            "removeValue": [Function],
                                            "reset": [Function],
                                            "setError": [Function],
                                            "setSuccess": [Function],
                                            "setTouched": [Function],
                                            "setValue": [Function],
                                            "setWarning": [Function],
                                            "success": undefined,
                                            "swapValues": [Function],
                                            "touched": undefined,
                                            "validate": [Function],
                                            "validatingField": [Function],
                                            "value": undefined,
                                            "warning": undefined,
                                          }
                                        }
                                        id="username"
                                        maxLength={30}
                                        placeholder="John Doe"
                                        required={true}
                                      >
                                        <input
                                          id="username"
                                          maxLength={30}
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          placeholder="John Doe"
                                          required={true}
                                          value=""
                                        />
                                      </TextWrapper>
                                    </Field>
                                  </ConnectedField>
                                  <label
                                    htmlFor="username"
                                  >
                                    UserName
                                  </label>
                                </div>
                              </StyledComponent>
                            </styled.div>
                            <styled.div>
                              <StyledComponent
                                forwardedComponent={
                                  Object {
                                    "$$typeof": Symbol(react.forward_ref),
                                    "attrs": Array [],
                                    "componentStyle": ComponentStyle {
                                      "componentId": "sc-bwzfXH",
                                      "isStatic": false,
                                      "lastClassName": "fsDvwN",
                                      "rules": Array [
                                        "
  display: block;
  padding-top: ",
                                        "21.25px",
                                        ";
  margin-bottom: ",
                                        "20px",
                                        ";
  position: relative;

  & input,
  & textarea {
    display: block;
    background-color: transparent;
    color: ",
                                        "rgba(0, 0, 0, 0.87)",
                                        ";
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.26);
    outline: 0;
    width: 100%;
    padding: 0;
    box-shadow: none;
    border-radius: 0;
    font-size: ",
                                        "17px",
                                        ";
    font-family: inherit;
    line-height: inherit;
    background-image: none;
    text-indent: 0.5em;
  }

  & > label {
    position: absolute;
    top: 0;
    display: block;
    width: 100%;
    color: ",
                                        "rgba(0, 0, 0, 0.54)",
                                        ";
    font-size: 12px;
    font-weight: 400;
    line-height: 15px;
    overflow-x: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    text-align: left;
  }

  & input[type=\\"time\\"] {
    cursor: pointer;
  }

  /*
   * autocomplete styles in webkit browsers
   * https://css-tricks.com/snippets/css/change-autocomplete-styles-webkit-browsers/
  */
  & input:-webkit-autofill,
  & input:-webkit-autofill:hover,
  & input:-webkit-autofill:focus,
  & input:-webkit-autofill:active {
    box-shadow: 0 0 0px 10px ",
                                        [Function],
                                        "
      inset;
    -webkit-text-fill-color: ",
                                        [Function],
                                        ";
  }

  &.tms-textfield--table input {
    border-bottom: none;
  }
",
                                      ],
                                    },
                                    "displayName": "styled.div",
                                    "foldedComponentIds": Array [],
                                    "render": [Function],
                                    "styledComponentId": "sc-bwzfXH",
                                    "target": "div",
                                    "toString": [Function],
                                    "warnTooManyClasses": [Function],
                                    "withComponent": [Function],
                                  }
                                }
                                forwardedRef={null}
                              >
                                <div
                                  className="sc-bwzfXH fsDvwN"
                                >
                                  <ConnectedField
                                    field="password"
                                    id="password"
                                    maxLength={30}
                                    placeholder="Your password"
                                    required={true}
                                    type="password"
                                  >
                                    <Field
                                      component={[Function]}
                                      field="password"
                                      id="password"
                                      maxLength={30}
                                      placeholder="Your password"
                                      pure={true}
                                      required={true}
                                      type="password"
                                    >
                                      <TextWrapper
                                        fieldApi={
                                          Object {
                                            "addValue": [Function],
                                            "asyncValidate": [Function],
                                            "doneValidatingField": [Function],
                                            "error": undefined,
                                            "fieldName": "password",
                                            "preValidate": [Function],
                                            "removeValue": [Function],
                                            "reset": [Function],
                                            "setError": [Function],
                                            "setSuccess": [Function],
                                            "setTouched": [Function],
                                            "setValue": [Function],
                                            "setWarning": [Function],
                                            "success": undefined,
                                            "swapValues": [Function],
                                            "touched": undefined,
                                            "validate": [Function],
                                            "validatingField": [Function],
                                            "value": undefined,
                                            "warning": undefined,
                                          }
                                        }
                                        id="password"
                                        maxLength={30}
                                        placeholder="Your password"
                                        required={true}
                                        type="password"
                                      >
                                        <input
                                          id="password"
                                          maxLength={30}
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          placeholder="Your password"
                                          required={true}
                                          type="password"
                                          value=""
                                        />
                                      </TextWrapper>
                                    </Field>
                                  </ConnectedField>
                                  <label
                                    htmlFor="password"
                                  >
                                    Password
                                  </label>
                                </div>
                              </StyledComponent>
                            </styled.div>
                            <button
                              className="tms-btn primary block"
                            >
                              SIGN IN
                            </button>
                          </form>
                        </div>
                      </StyledComponent>
                    </styled.div>
                  </Form>
                </Connect(Form)>
              </ReactForm>
            </div>
          </StyledComponent>
        </styled.div>
      </main>
    </StyledComponent>
  </styled.main>
</Component>
`;
